### PR TITLE
Add Theme provider

### PR DESCRIPTION
**Pull request by cvarjao**
# Summary of Changes

- Add Theme provider
- Fixed lintstagedrc to run using npx so that there is no need to have dependencies installed globally (e.: json, tsx, typescript)
- Removed import aliasing (`components`)
- Added `typecheck` script so that we can use for typescript validation (`npm run typecheck`)

# Related Issues

Please reference here any issue #'s that are relevant to this PR, or simply enter "N/A" if this PR does not relate to any existing issues.

# Pull Request Checklist

This is just a reminder about the most common mistakes. Please make sure that you tick all _appropriate_ boxes. But please read our [contribution guide](../CONTRIBUTING.md) at least once; it will save you a few review cycles!

If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing to do.

- [X] All commits contain a DCO `Signed-off-by` line (we use the [DCO GitHub app](https://github.com/apps/dco) to enforce this).
- [X] Updated LICENSE-3RD-PARTY.md for any added dependencies or vendored components.
- [X] Run prettier: `npm run style-format`
- [ ] Updated **documentation** for changed code and new or modified features.

If you have _any_ questions to _any_ of the points above, just **submit and ask**! This checklist is here to _help_ you, not to deter you from contributing!

_PR template adapted from the Python attrs project._

### CODE REVIEW

**Comment by cvarjao** [comment]
@JamesKEbert , @TelegramSam, would any of you have a chance to review this initial PR? I am holding off on the structural changes for the next PR

File: .lintstagedrc.json [hunk 1]
```diff
@@ -1,3 +1,3 @@
 {
-  "**/*.{js,jsx,ts,tsx}": ["npm run lint -- --fix", "npm run prettier -- --write"]
+  "**/*.{js,jsx,ts,tsx}": ["npm run lint -- --fix", "npx prettier --write"]
```
Comment (jleach) [inline .lintstagedrc.json:2]: Why this change only for prettier and not lint also?

File: App.tsx [hunk 2]
```diff
@@ -16,14 +16,20 @@ import SplashScreen from 'react-native-splash-screen'
 import Toast from 'react-native-toast-message'
 
 import ConnectionModal from './App/components/modals/ConnectionModal'
+import ErrorModal from './App/components/modals/ErrorModal'
 import toastConfig from './App/components/toast/ToastConfig'
-import { initStoredLanguage } from './App/localization'
+import { initLanguages, initStoredLanguage, defaultTranslationResources } from './App/localization'
 import RootStack from './App/navigators/RootStack'
+import { pages as defaultOnboardingPages } from './App/screens/OnboardingPages'
+import { default as defaultSplashScreen } from './App/screens/Splash'
```
Comment (jleach) [inline App.tsx]: Do these need to be changed? Typically when we `export default Splash` you just need to `import Splash from ...`. Is it because you need a specific name?

File: App/components/buttons/Button.tsx [hunk 3]
```diff
@@ -1,14 +1,14 @@
...
-import { Buttons, heavyOpacity } from '../../theme'
+import { useThemeContext } from '../../utils/themeContext'
 
 export enum ButtonType {
   Primary,
   Secondary,
 }
 
-interface ButtonProps {
+export interface ButtonProps {
```
Comment (jleach) [inline App/components/buttons/Button.tsx]: Do these get imported elsewhere in the app now? I didn't think they needed to be exported.

File: App/components/inputs/SingleSelectBlock.tsx [hunk 4]
```diff
@@ -2,25 +2,9 @@ import React, { useState } from 'react'
 import { StyleSheet, TouchableOpacity, View } from 'react-native'
 import Icon from 'react-native-vector-icons/MaterialIcons'
 
-import { borderRadius, ColorPallet } from '../../theme'
+import { Text } from '../'
```
Comment (jleach) [inline App/components/inputs/SingleSelectBlock.tsx]: How do you feel about using the full path rather than `../`? If I read this line and needed to change the Text component I'd hit Command-Shift-P in VSCode and open the file by name. With `../` I may have to lookup the file name or use the navigator if I didn't know it. I think it makes the code more readable.

File: App/screens/OnboardingPages.tsx [hunk 5]
```diff
@@ -6,76 +6,88 @@ import CredentialList from '../assets/img/credential-list.svg'
...
-    marginRight: 20,
-    marginBottom: 10,
-  },
-  icon: {
-    marginRight: 10,
-  },
-})
-
-const customPages = (onTutorialCompleted: GenericFn) => {
+export function createStyle(theme: Theme) {
```
Comment (jleach) [inline App/screens/OnboardingPages.tsx]: If you're refactoring in the future feel free to update these to the arrow notation `() => {}`. It's more uniform.

File: App/screens/ProofRequest.tsx [hunk 6]
```diff
@@ -88,15 +86,23 @@ const ProofRequest: React.FC<ProofRequestProps> = ({ navigation, route }) => {
...
+        throw error
       }
-      setCredentials(creds)
-      setAttributeCredentials(Object.entries(creds?.requestedAttributes || {}))
     }
 
-    updateRetrievedCredentials(proof).catch(() => {
+    updateRetrievedCredentials(proof).catch((err) => {
+      // eslint-disable-next-line no-console
+      console.log(`ERROR2:`, JSON.stringify(err))
```
Comment (jleach) [inline App/screens/ProofRequest.tsx]: Able to remove this `console.log`?

File: App/utils/configurationContext.tsx [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import { createContext, useContext, FC } from 'react'
+
+import { pages } from '../screens/OnboardingPages'
+
+export interface ConfigurationContext {
+  onboarding: {
+    pages: typeof pages
+  }
+  splash: FC
```
Comment (jleach) [inline App/utils/configurationContext.tsx]: I think it would be more readable to use `React.FC` here.

File: __tests__/components/AvatarView.test.tsx [hunk 8]
```diff
@@ -2,9 +2,12 @@ import { render } from '@testing-library/react-native'
 import React from 'react'
 
 import AvatarView from '../../App/components/misc/AvatarView'
+import * as themeContext from '../../App/utils/themeContext' // note we're importing with a * to import all the exports
+import { defaultTheme } from '../../App/utils/themeContext'
 
 describe('AvatarView Component', () => {
   test('Renders correctly', async () => {
+    jest.spyOn(themeContext, 'useThemeContext').mockImplementation(() => defaultTheme)
```
Comment (jleach) [inline __tests__/components/AvatarView.test.tsx:10]: Would it be easier to add a global mock in `__mock__`?

**Review by jleach — CHANGES_REQUESTED** [review]
Most of my comments are not deal breakers for merging but I'd like to see the reply before approving and give others a chance to review so I marked my review as Request Changes.

While reviewing I was wondering if ti would be better to keep vars that don't change like `styles` out of the FC? I've always just put things that change inside and things that don't outside. Then you could just do something like `style.foo = { ...style.foo, ...{color: 'red}}` to override or use `style.compose()` to override inside the FC itself.

File: .lintstagedrc.json (hunk 1 above)
Comment (cvarjao) [reply .lintstagedrc.json:2]: That is the only one that was broken when I am running on my windows laptop. Didn't touch anything else that was already working

File: App.tsx (hunk 2 above)
Comment (cvarjao) [reply App.tsx]: Either one works. No specific reason.  Happy to update

File: App/components/inputs/SingleSelectBlock.tsx (hunk 4 above)
Comment (cvarjao) [reply App/components/inputs/SingleSelectBlock.tsx]: I am not to opinionated. The only problem is that aliases break when using as a library

File: App/screens/ProofRequest.tsx (hunk 6 above)
Comment (cvarjao) [reply App/screens/ProofRequest.tsx]: Yes! left over.

File: App.tsx (hunk 2 above)
Comment (JamesKEbert) [reply App.tsx]: +1 on this thought

File: App/components/inputs/SingleSelectBlock.tsx (hunk 4 above)
Comment (JamesKEbert) [reply App/components/inputs/SingleSelectBlock.tsx]: So @jleach you're proposing `'../TextInput'` (which I am a fan of), but to confirm would this cause problems from a library perspective @cvarjao ?

File: App/components/listItems/NotificationListItem.tsx [hunk 9]
```diff
@@ -7,12 +7,11 @@ import { useTranslation } from 'react-i18next'
...
 import Icon from 'react-native-vector-icons/MaterialIcons'
 
-import { TextTheme, ColorPallet } from '../../theme'
 import { GenericFn } from '../../types/fn'
 import { HomeStackParams, NotificationStackParams, Screens, Stacks } from '../../types/navigators'
 import { parsedSchema } from '../../utils/helpers'
-
-import Button, { ButtonType } from 'components/buttons/Button'
+import { useThemeContext } from '../../utils/themeContext'
+import Button, { ButtonType } from '..//buttons/Button'
```
Comment (JamesKEbert) [inline App/components/listItems/NotificationListItem.tsx]:
```suggestion
import Button, { ButtonType } from '../buttons/Button'
```
One too many slashes?

File: App/types/navigators.ts [hunk 10]
```diff
@@ -72,6 +72,7 @@ export type CredentialStackParams = {
 export type HomeStackParams = {
   [Screens.Home]: undefined
   [Screens.Notifications]: undefined
+  [Screens.ProofRequestAttributeDetails]: { proofId: string; attributeName: string }
```
Comment (JamesKEbert) [inline App/types/navigators.ts:75]: Maybe I'm missing something but this seems to be duplicated with line 90 below?

**Review by JamesKEbert — DISMISSED** [review]
This looks good! I left pretty minor comments. Good work, thanks for working on this! :)

File: App/components/buttons/Button.tsx (hunk 3 above)
Comment (cvarjao) [reply App/components/buttons/Button.tsx]: Not for this PR, it will be needed eventually for the bifold as framework, but I guess we also need to have a conversation about how to implement models/interfaces that downstream projects can use. I will remove the export for now.

File: App/types/navigators.ts (hunk 10 above)
Comment (cvarjao) [reply App/types/navigators.ts:75]: Each stack have their own type, but they are very similar. There is probably an opportunity to refactor the code. But if it is alright, I will leave that for another work.

File: __tests__/components/AvatarView.test.tsx (hunk 8 above)
Comment (cvarjao) [reply __tests__/components/AvatarView.test.tsx:10]: I didn't want to mock the whole `themeContext.ts` component, only that particular function. I am not sure how to that using `__mock__`. I previously had in a [setup file](https://jestjs.io/docs/configuration#setupfiles-array) , but I didn't want to introduce that just yet on this PR

Comment (cvarjao) [reply __tests__/components/AvatarView.test.tsx:10]: Also, `__tests__` are currently being ignored from typescript validation because there are a bunch of false positive errors. I think we will need to review and address  those those errors (possibly refactor) as another work.

File: App/components/inputs/SingleSelectBlock.tsx (hunk 4 above)
Comment (amanji) [reply App/components/inputs/SingleSelectBlock.tsx]: I agree with being more explicit here for developer experience

**Comment by amanji** [comment]
Wondering if we can further split this out into separate PRs, one for the config provider and one for the theme provider? Unless the changes need to be done together.

File: App/components/buttons/HeaderRight.tsx [hunk 11]
```diff
@@ -12,27 +12,26 @@ interface HeaderButtonProps {
...
-  icon: {
-    transform: [{ rotate: '180deg' }],
-  },
-})
-
 const HeaderRight: React.FC<HeaderButtonProps> = ({ title, testID, accessibilityLabel, onPress, disabled = false }) => {
   const accessible = accessibilityLabel && accessibilityLabel !== '' ? true : false
-
+  const { ColorPallet, TextTheme } = useThemeContext()
+  const style = StyleSheet.create({
```
Comment (amanji) [inline App/components/buttons/HeaderRight.tsx:18]: While I understand why we need to move the style configs into the component function I'm wondering if there is a better way to do this in order to keep function bodies lean and easier to read.

File: App/components/inputs/CheckBoxRow.tsx [hunk 12]
```diff
@@ -2,31 +2,31 @@ import React from 'react'
 import { View, StyleSheet, TouchableOpacity, Text } from 'react-native'
 import Icon from 'react-native-vector-icons/MaterialIcons'
 
-import { ColorPallet, TextTheme } from '../../theme'
+import { useThemeContext } from '../../utils/themeContext'
 
-interface Props {
+export interface Props {
```
Comment (amanji) [inline App/components/inputs/CheckBoxRow.tsx]: Export introduced

File: App/components/misc/QRScanner.tsx [hunk 13]
```diff
@@ -5,48 +5,17 @@ import { useWindowDimensions, Vibration, View, StyleSheet, Text } from 'react-na
...
 import Icon from 'react-native-vector-icons/MaterialIcons'
 
-import { ColorPallet, TextTheme } from '../../theme'
-
-import QRScannerClose from 'components/misc/QRScannerClose'
-import QRScannerTorch from 'components/misc/QRScannerTorch'
-import { QrCodeScanError } from 'types/error'
+import { QrCodeScanError } from '../../types/error'
+import { useThemeContext } from '../../utils/themeContext'
+import QRScannerClose from '../misc/QRScannerClose'
```
Comment (amanji) [inline App/components/misc/QRScanner.tsx]: These components are in the same directory so it may just be cleaner to use `./QRScannerClose`

File: App/components/misc/QRScannerTorch.tsx [hunk 14]
```diff
@@ -2,34 +2,38 @@ import React from 'react'
...
-    borderRadius: 24,
-    marginBottom: 50,
-  },
-  icon: {
-    marginLeft: 2,
-    marginTop: 2,
-  },
-})
-
+function createStyles({ ColorPallet }: Theme) {
```
Comment (amanji) [inline App/components/misc/QRScannerTorch.tsx:13]: I like the idea used here to add styling without cluttering up the component function in reference to my other comment.

File: App/components/buttons/HeaderRight.tsx (hunk 11 above)
Comment (amanji) [reply App/components/buttons/HeaderRight.tsx:18]: Saw @jleach comment above and I agree with the sentiment of trying to keep things clean.

File: App/types/navigators.ts (hunk 10 above)
Comment (amanji) [reply App/types/navigators.ts:75]: This line should be removed. That screen should only be navigated to from a proof request details screen and hence is grouped as part of a different set of stack parameters. Having it as part of `HomeStackParams` is confusing.

File: App/components/inputs/CheckBoxRow.tsx (hunk 12 above)
Comment (amanji) [reply App/components/inputs/CheckBoxRow.tsx]: While we're at it we've been moving away from using the generic term `Props` and prefixing with the component name so this may be a good opportunity to rename to `CheckBoxRowProps`

File: App.tsx (hunk 2 above)
Comment (jleach) [reply App.tsx]: @cvarjao Let's roll with the simplified notation:`import xxx from 'yyy'`

File: __tests__/components/AvatarView.test.tsx (hunk 8 above)
Comment (jleach) [reply __tests__/components/AvatarView.test.tsx:10]: Ok. We can leave it for now and I can show you some mocking tricks. I think you'll create a file in `__mock__` and use  `jest.requireActual()` or something then just mock what you want - not the entire thing.

**Review by jleach — DISMISSED** [review]

**Review by amanji — DISMISSED** [review]

**Review by jleach — APPROVED** [review]

**Review by amanji — APPROVED** [review]
